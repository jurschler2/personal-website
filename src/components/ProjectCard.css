/** 
 * Styling for the ProjectCard component:
 */

.project-description-container {
  margin-top: 2%;
  margin-bottom: 5%;
} 

.project-description {
  display: flex;
  width: 80%;
  margin-left: 10%;
  text-align: left;
  font-size: small;
}

.project-title {
  width: 80%;
  margin-left: 10%;
  margin-bottom: 1%;
  text-align: center;
}

.project-img-container {
  display: flex;
  width: 80%;
  margin-left: 10%;
  align-items: center;
}

.carousel-item > img { 
  width: 100%; 
  border-radius: 5px;
}

.carousel-control-prev,.carousel-control-next {
  cursor:pointer;
}

.carousel-caption {
  font-size: 3vw;
  letter-spacing: .5px;
    text-shadow: -1px 1px 0 #000,
				  1px 1px 0 #000,
				 1px -1px 0 #000,
				-1px -1px 0 #000;
}

 

 @media screen and (min-width: 760px) {
  .project-container {
    display: inline-block;
    width: 100%;
    padding: 1%;
  }
 
  .project-description-container {
    display: inline-block;
    width: 60%;
    text-align: left;
  }
  
  .project-title {
    text-align: center;
  }

  .project-description {
    display: inline-block;
    width: 80%;
    text-align: left;
    font-size: small;
  }


  .project-img-container {
    display: inline-block;
    text-align: center;
    width: 30%;
    vertical-align: top;
    position: relative;
    
  }

  .carousel-item > img { 
    width: 100%; 
    border-radius: 5px;
  }

  .carousel-control-prev,.carousel-control-next {
    cursor:pointer;
  }

  .carousel-caption {
    font-size: 1vw;
    letter-spacing: 1px;
      text-shadow: -1px 1px 0 #000,
            1px 1px 0 #000,
           1px -1px 0 #000,
          -1px -1px 0 #000;
  }

}