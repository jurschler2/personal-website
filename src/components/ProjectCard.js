import React, {useState} from "react";
import {
  Carousel,
  CarouselItem,
  CarouselControl,
  CarouselIndicators
} from 'reactstrap';
import { Card, Row, Col } from 'react-bootstrap';

/**
 *  DESCRIPTION:
 *  PROPS: title, description, img
 *  FLOW: App => Routes => Portfolio => ProjectCard
 *  PARENT: Portfolio
 *  CHILDREN: ProjectCard
 */

function ProjectCard({ title, description, gallery, demo, link, tech }) {

  const [activeIndex, setActiveIndex] = useState(0);
  const [animating, setAnimating] = useState(false);

  const next = () => {
    if (animating) return;
    const nextIndex = activeIndex === gallery.length - 1 ? 0 : activeIndex + 1;
    setActiveIndex(nextIndex);
  }

  const previous = () => {
    if (animating) return;
    const nextIndex = activeIndex === 0 ? gallery.length - 1 : activeIndex - 1;
    setActiveIndex(nextIndex);
  }

  const goToIndex = (newIndex) => {
    if (animating) return;
    setActiveIndex(newIndex);
  }

  const slides = gallery.map((item) => {
    return (
      <CarouselItem
        onExiting={() => setAnimating(true)}
        onExited={() => setAnimating(false)}
        key={item.img}
      >
        <img className="projectImage" src={item.img} alt={item.altText}/>
      </CarouselItem>
    );
  });

  const createTechTags = (tech, title) => {
    let tags = [];
    for (let t of tech) {
      if (tags.length === tech.length - 1) {
        tags.push(<span key={`${title}-${t}`} className="projectTechStack">{t}</span>);
      } else {
        tags.push(
          <span key={`${title}-${t}`} className="projectTechStack">{t} <span className="projectTechStackSep">| </span></span>
        );
      }
    }
    return tags;
  }

  const techStack = createTechTags(tech, title);

  return (
  <Card className="project-container col-12" data-aos="fade-right">
    <Row>
      <Col md={12} lg={12}>
        <div className="projectImageContainer">
          <Carousel
            activeIndex={activeIndex}
            next={next}
            previous={previous}
            >
            <CarouselIndicators items={gallery} activeIndex={activeIndex} onClickHandler={goToIndex} />
              {slides}
            <CarouselControl direction="prev" directionText="Previous" onClickHandler={previous} />
            <CarouselControl direction="next" directionText="Next" onClickHandler={next} />
          </Carousel>    
        </div>
      </Col>
      <Col md={12} lg={12} className="projectDetails">
        <div>
            <h3 className="project-title"><strong>{title}</strong></h3>
            <p className="projectTechStack">{techStack}</p>
            <div className="project-description">
  <p>{description} Check it out on <a href={`${link}`}>Github</a>. Check out the <a href={`${demo}`}>demo</a>.</p>
          </div>
        </div>
      </Col>
    </Row>
  </Card>
  )
}

export default ProjectCard;