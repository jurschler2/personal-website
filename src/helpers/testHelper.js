import React from "react";
import { render } from "@testing-library/react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";

export function renderWithRouter(children, initialPath = "") {
  const history = createMemoryHistory();
  if (initialPath) history.push(initialPath);
  return render(<Router history={history}>{children}</Router>);
}
