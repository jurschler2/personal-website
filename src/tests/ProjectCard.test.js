import React from "react";
import ProjectCard from "../components/ProjectCard";
import {renderWithRouter} from "../helpers/testHelper"

it("renders without crashing", () => {
  renderWithRouter(<ProjectCard description="description" title="title" img="img" />);
});

it("matches snapshot", () => {
  const { asFragment } = renderWithRouter(<ProjectCard description="description" title="title" img="img" />);
  expect(asFragment()).toMatchSnapshot();
});