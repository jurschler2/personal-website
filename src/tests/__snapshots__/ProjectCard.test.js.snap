// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <span>
      title
    </span>
    This is the description and this is the img.
  </div>
</DocumentFragment>
`;
